Fixed #36458 Move focus to admin widget popup window after button click

XChaitanyaX opened the pull request:
#### Trac ticket number
<!-- Replace XXXXX with the corresponding Trac ticket number, or delete the line and write "N/A" if this is a trivial PR. -->

ticket-36458

#### Branch description
The proposed change ensures that when a popup widget is triggered, keyboard focus is programmatically moved to the desired element (current date for the date picker and the 'Now' element in the time picker) within the popup.

#### Checklist
- [x] This PR targets the `main` branch. <!-- Backports will be evaluated and done by mergers, when necessary. -->
- [x] The commit message is written in past tense, mentions the ticket number, and ends with a period.
- [x] I have checked the "Has patch" ticket flag in the Trac system.
- [ ] I have added or updated relevant tests.
- [ ] I have added or updated relevant docs, including release notes if applicable.
- [ ] I have attached screenshots in both light and dark modes for any UI changes.

XChaitanyaX committed:
Fixed #36458 Move focus to admin widget popup window after button click

Antoliny0919 commented on django/contrib/admin/static/admin/js/admin/DateTimeShortcuts.js:
```diff
@@ -179,6 +181,36 @@
             });
 
+            // Handle tab navigation within the popup (after content is created)
+            clock_box.addEventListener('keydown', function(event) {
```
The changes added are great, but it doesn't seem like screen readers were taken into consideration.
When using a screen reader, users can navigate specific elements using the Tab key.
The internal elements of the dialogs provided by the admin (such as Date and Time pickers) are implemented as `<a>` tags, so they are navigable using the Tab key while a screen reader is active.
The current code added seems to be designed for navigation with a regular keyboard.

Antoliny0919 commented on django/contrib/admin/static/admin/js/admin/DateTimeShortcuts.js:
```diff
@@ -376,12 +454,58 @@
             document.removeEventListener('click', DateTimeShortcuts.dismissCalendarFunc[num]);
         },
+        // Helper function to focus the next element in the tab order
+        focusNextElement: function(currentElement) {
+            // Get all tabbable elements in the document, excluding hidden popup elements
+            const allTabbableElements = document.querySelectorAll(
```
I don’t think most of the changes are designed with screen readers in mind.
This is the same as my comment above.

XChaitanyaX replied:
Sure thing, previously the Tab key is not even working as intended to (the users can still navigate but in a very awkward way i.e. to go through all the elements and then the time and date elements are focused which is annoying for the specific types of users) as the popup elements are not even focused when entered. And the Tab is now working (but not focusing to the current date or time as you mentioned below). Thank you for the comments, I will try to fix them.

Antoliny0919 reviewed (commented):
Thank you for picking up this ticket @XChaitanyaX ⭐️

I've added a couple of comments inline below.

Antoliny0919 commented:
The ticket you picked aims to improve accessibility for screen readers.

## The goals of this PR are as follows.

When a user using a screen reader clicks on the datepicker or timepicker provided by the admin, focus should be immediately set to the dialog that appears.

## Where should the focus be set?

In my opinion, since the Date and Time elements represent time, it would be best if the initial focus (the first element accessed) is set to the element that indicates the current time. This is just my personal view, but both W3C and the `<input type="date">` element set the focus directly to the element representing the current time when accessed with a screen reader.

- Date

The element to be focused when pressing Enter in a screen reader...

<img width="1281" height="752" alt="date_screenreader_example" src="https://github.com/user-attachments/assets/2beef5d5-92e7-418b-9cdd-23ea26af0810" />

- Time

The element to be focused when pressing Enter in a screen reader...

<img width="1124" height="696" alt="time_screenreader_example" src="https://github.com/user-attachments/assets/3129a362-3bd6-4f74-b053-f6cb489c46d5" />

XChaitanyaX committed:
Fixed #36458 Move focus to current date and time for admin date, time elements

XChaitanyaX commented:
https://github.com/user-attachments/assets/d46004dd-367d-428e-a5d6-7669f1f05883

The video shows a small lag (which actually is not there when I tested it) when the clock-box is clicked (also at the calendar-box focusing the current date) to focus the Now element.
Now the elements are properly focused.

XChaitanyaX committed:
Enhance accessibility for date and time pickers with ARIA attributes and add keyboard navigation to the admin widgets

XChaitanyaX committed:
remove unnecessary try catch blocks and indents

XChaitanyaX commented:
https://github.com/user-attachments/assets/d439e48a-d1a0-4278-8407-1cdbb4441e01

- I am using Narrator (a screen reader) in windows
- Is this good or somethings still need change

Antoliny0919 reviewed (commented):
Thank you @XChaitanyaX .

I worked on the date picker separately.

```diff --git a/django/contrib/admin/static/admin/js/admin/DateTimeShortcuts.js b/django/contrib/admin/static/admin/js/admin/DateTimeShortcuts.js
index 8168172a97..73802c26b9 100644
--- a/django/contrib/admin/static/admin/js/admin/DateTimeShortcuts.js
+++ b/django/contrib/admin/static/admin/js/admin/DateTimeShortcuts.js
@@ -245,11 +245,13 @@
             const cal_link = document.createElement('a');
             cal_link.href = '#';
             cal_link.id = DateTimeShortcuts.calendarLinkName + num;
-            cal_link.addEventListener('click', function(e) {
-                e.preventDefault();
-                // avoid triggering the document click handler to dismiss the calendar
-                e.stopPropagation();
-                DateTimeShortcuts.openCalendar(num);
+            ['click', 'keydown'].forEach((event) => {
+                cal_link.addEventListener(event, function(e) {
+                    e.preventDefault();
+                    // avoid triggering the document click handler to dismiss the calendar
+                    e.stopPropagation();
+                    DateTimeShortcuts.openCalendar(num, e);
+                });
             });
             quickElement(
                 'span', cal_link, '',
@@ -345,7 +347,7 @@
                 }
             });
         },
-        openCalendar: function(num) {
+        openCalendar: function(num, e) {
             const cal_box = document.getElementById(DateTimeShortcuts.calendarDivName1 + num);
             const cal_link = document.getElementById(DateTimeShortcuts.calendarLinkName + num);
             const inp = DateTimeShortcuts.calendarInputs[num];
@@ -376,6 +378,11 @@
             cal_box.style.top = Math.max(0, findPosY(cal_link) - 75) + 'px';
 
             cal_box.style.display = 'block';
+            if (e.key === 'Enter') {
+                const selected = cal_box.querySelector('td.selected a');
+                const focusElement = selected ? selected : cal_box.querySelector('td.today a');
+                focusElement.focus();
+            }
             document.addEventListener('click', DateTimeShortcuts.dismissCalendarFunc[num]);
         },
         dismissCalendar: function(num) {
```
When a screen reader user accesses the date picker element, the above code alone is sufficient to focus on a specific element (the selected date if one exists, or today if not), allowing them to freely navigate the dialogs.

I think this change is sufficient for improving the accessibility of the date picker in this ticket for now.

XChaitanyaX committed:
removed aira attributes, removed changes that doesn't belong to the ticket

XChaitanyaX commented:
### Changes done
- removed aria attributes.
- removed accessibility via keyboard where arrow keys are used to navigate in both the date and time pickers as this change is irrelevant to this ticket
- using the class of the selected date to find the element in the date picker dialog.

Antoliny0919 commented on django/contrib/admin/static/admin/js/admin/DateTimeShortcuts.js:
```diff
@@ -153,18 +156,27 @@
+            // Add ARIA attributes for better screen reader support
+            clock_box.setAttribute('role', 'dialog');
+            clock_box.setAttribute('aria-label', gettext('Choose a time'));
+            clock_box.setAttribute('aria-modal', 'true');
```
I agree that the aria attributes should be added. However, as you mentioned, this should be handled separately from this ticket.
Since the two widgets have a similar structure, I also think it might be better to handle them together for the sake of consistency, but hmm... I guess that’s something to think about later. 😄

Antoliny0919 commented:
@XChaitanyaX
If you’d like to have your work reviewed(fellow), please remove the "patch needs improvement" and "needs tests" flags from the ticket.
Feel free to mention me, and I’ll be happy to review it again.

Antoliny0919 commented:
++ When the picker is activated by pressing Enter without using a screen reader, the user still has to navigate through all the elements, just like a screen reader user would. However, I believe this improvement will address that issue for both cases.

XChaitanyaX commented:
> ++ When the picker is activated by pressing Enter without using a screen reader, the user still has to navigate through all the elements, just like a screen reader user would. However, I believe this improvement will address that issue for both cases.

Yes, it will fix for both the cases

- The accessibility can be improved even more if we add arrow keys to navigate elements in the pickers (it would be a great improvement in the date picker).
- if the current date is 1st and the user need to select the last date of the month, pressing the Tab key 30 times is annoying, if we have arrow key navigation in the date picker it would be a lot easier.

Antoliny0919 commented:
> Yes, it will fix for both the cases
> 
>     * The accessibility can be improved even more if we add arrow keys to navigate elements in the pickers (it would be a great improvement in the date picker).
> 
>     * if the current date is 1st and the user need to select the last date of the month, pressing the Tab key 30 times is annoying, if we have arrow key navigation in the date picker it would be a lot easier.

Yes, I think so too. The [W3C example date picker](https://www.w3.org/WAI/ARIA/apg/patterns/dialog-modal/examples/datepicker-dialog/) also provides this functionality.
However, this feature is also something that should be addressed in this [ticket](code.djangoproject.com/ticket/36459).

sarahboyce commented on django/contrib/admin/static/admin/js/admin/DateTimeShortcuts.js:
```diff
@@ -125,8 +127,7 @@
 
             quickElement(
                 'span', clock_link, '',
-                'class', 'clock-icon',
-                'title', gettext('Choose a Time')
+                'class', 'clock-icon'
```
Note that if possible, only include changes that are relevant for the ticket 
Other fixes should be included with other associated tickets (in a separate commit) :+1:

sarahboyce reviewed (commented):
Thank you for the PR @XChaitanyaX :star: 

Note that we will need tests for the changes, this could either be a selenium test or a javascript test in `js_tests/admin/DateTimeShortcuts.test.js`

I have tested and the focus does go in the popup window of the datepicker, however:
- when I have tabbed through the options, I should loop within the datepicker. Instead I then tab through the top of the page
- when I hit "esc" I should be focused on the trigger button (and can continue to tab from their) but instead I am at the top of the page

See: https://www.w3.org/WAI/ARIA/apg/patterns/dialog-modal/#keyboardinteraction

XChaitanyaX committed:
fixed incorrect navigation while using Tab and esc keys

XChaitanyaX commented:
@Antoliny0919 
Any idea on how to write tests for the changes??

Antoliny0919 commented:
> Thank you for the PR @XChaitanyaX ⭐
> 
> Note that we will need tests for the changes, this could either be a selenium test or a javascript test in `js_tests/admin/DateTimeShortcuts.test.js`
> 
> I have tested and the focus does go in the popup window of the datepicker, however:
> 
>     * when I have tabbed through the options, I should loop within the datepicker. Instead I then tab through the top of the page
> 
>     * when I hit "esc" I should be focused on the trigger button (and can continue to tab from their) but instead I am at the top of the page
> 
> 
> See: https://www.w3.org/WAI/ARIA/apg/patterns/dialog-modal/#keyboardinteraction

Should the tab navigation within the datepicker also be part of this work?
I thought handling tab navigation inside the dialog should be addressed in the DatePicker accessibility improvement ticket.
But if it is meant to be included here, I’ll make sure to review the focus navigation as well 🥰

sarahboyce commented:
> Should the tab navigation within the datepicker also be part of this work?

The same is true of the clock widget. I think this ticket should ideally be scoped for keyboard navigation of widget popups :+1:

Antoliny0919 commented:
So, the goals we should aim to achieve in this PR are as follows:

- Assign focus to the appropriate element when pressing Enter on the `DateTime` picker icon.
As mentioned before, for the `Date` field, if no date is selected, the focus should default to today’s date. If a date is selected, the focus should move to that date.
For the `Time` field, focusing on “now” seems to be the most appropriate behavior.

- Enable keyboard navigation within the `DateTime` dialog.
Similar to the [W3C example for date pickers](https://www.w3.org/WAI/ARIA/apg/patterns/dialog-modal/examples/datepicker-dialog/), users should be able to navigate between day elements using the arrow keys.
This is important because the focus should not be limited to just date selection, users must also be able to access other interactive elements like "Yesterday," "Tomorrow," "Close," etc.
Additionally, as Sarah suggested, we need to consider what happens when the user reaches the last focusable element.
In the W3C example, focus moves to the next or previous month in such cases.

Thinking about it, implementing all of these features would likely require a significant amount of Javascript and test coverage.
As Sarah previously suggested, switching to a native [`input type="date"`](https://developer.mozilla.org/ko/docs/Web/HTML/Reference/Elements/input/date) element instead of maintaining the current widget could reduce the maintenance burden, and at the same time, naturally improve accessibility.
That makes suggestion feel even more appealing.

I believe we need a clear direction for what we want to achieve in this PR.
In my opinion, we should discuss this further with the accessibility team.

XChaitanyaX commented:
@thibaudcolas Any update by the accessibility team?

thibaudcolas commented on django/contrib/admin/static/admin/js/admin/DateTimeShortcuts.js:
```diff
@@ -205,10 +211,40 @@
             clock_box.style.display = 'block';
+            
+            // Add focus loop for keyboard navigation
+            DateTimeShortcuts.addFocusLoop(clock_box);
```
I’d prefer we use the terminology of "focus trap" which is industry-standard.

thibaudcolas commented on django/contrib/admin/static/admin/js/admin/DateTimeShortcuts.js:
```diff
@@ -205,10 +211,40 @@
+            if (DateTimeShortcuts.lastFocusedElement) {
+                DateTimeShortcuts.lastFocusedElement.focus();
+                DateTimeShortcuts.lastFocusedElement = null;
+            }
```
Rather than going custom for this, I think we should leverage `aria-controls`. Add the attribute to the time picker’s toggle button, then we can do:

```suggestion
            const clock_box = document.getElementById(DateTimeShortcuts.clockDivName + num);
            clock_box.style.display = 'none';
            const trigger = document.querySelector(`[aria-controls="${clock_box.id}"]`);
            trigger.focus();
```

thibaudcolas commented on django/contrib/admin/static/admin/js/admin/DateTimeShortcuts.js:
```diff
@@ -376,12 +420,110 @@
             document.removeEventListener('click', DateTimeShortcuts.dismissCalendarFunc[num]);
         },
+        // Add focus loop to keep focus within the popup even after reaching the last focusable element
+        addFocusLoop: function(container) {
```
I think we should try to change this to use the `inert` HTML attribute so we don’t reinvent our own focus trapping.

thibaudcolas reviewed (changes requested):
Tested with keyboard support and macOS VoiceOver - this is a marked improvement but I think the implementation is way more complex than it needs to be :/ And prone to implementation bugs / nuances that we’d be better off not having to navigate.

- The "pick element to focus" logic really could just be a well-crafted `querySelector` one-liner. There’s no need for it to be so complex.
- The "last focused element" logic is also unnecessary, really all we need to know is for a given widget’s chooser dialog which is its triggering element, which the `aria-controls` attribute is for.
- We shouldn’t have to add our own event handling for focus trapping. There are better techniques available today, with plenty good enough browser support.

I expect those implementation changes will lead to better support, less code, and also tests that are much easier to write. No need to test a lot of branching logic that never happens in real-world apps, no need to write our own focus trap testing if we use HTML attributes / elements.

XChaitanyaX committed:
clock thing

jacobtylerwalls commented:
Thanks for all the progress and reviews 💜 

@Skyiesac is continuing in #21042, reviews there are welcome.
